File: why_LeetCode/Day13/C++/Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
// 时间复杂度 O(m+n) 空间复杂度O(1)
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode head(-1);//头结点
        ListNode* pre=&head;
        int c=0;
        for(ListNode *pa=l1,*pb=l2; pa!=NULL||pb!=NULL;
            pa=pa==NULL?NULL:pa->next,pb=pb==NULL?NULL:pb->next,pre=pre->next)
            {
                int a=pa==NULL?0:pa->val;
                int b=pb==NULL?0:pb->val;
                int val=(a+b+c)%10;
                c=(a+b+c)/10;

                pre->next=new ListNode(val);
            }
        if(c>0)
            pre->next=new ListNode(c);
        return head.next;
    }
};

File: why_LeetCode/Day13/Python/Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        head=ListNode(-1)
        pre=head
        c=0
        while c or l1 or l2:
            c,val=divmod(c+(l1.val if l1 else 0) + (l2.val if l2 else 0),10)
            pre.next=ListNode(val)
            pre=pre.next
            l1=l1.next if l1 else None
            l2=l2.next if l2 else None
        return head.next